realsense/plugin.cpp

```diff
@@ -0,0 +1,172 @@
+static constexpr int IMAGE_HEIGHT = 480;
+static constexpr int FPS = 30;
+static constexpr int GYRO_RATE = 400; // 200 or 400
+static constexpr int ACCEL_RATE = 250; // 63 or 250
+
+typedef struct {
+	cv::Mat* img0;
+	cv::Mat* img1;
+	cv::Mat* rgb;
+	cv::Mat* depth;
+	int iteration;
+} cam_type;
+
+typedef struct {
+	rs2_vector* accel_data;
+	int iteration;
+} accel_type;
+
+class realsense : public plugin {
+public:
+	realsense(std::string name_, phonebook *pb_)
+	: plugin{name_, pb_}
+	, sb{pb->lookup_impl<switchboard>()}
+	, _m_imu_cam{sb->publish<imu_cam_type>("imu_cam")}
+	, _m_rgb_depth{sb->publish<rgb_depth_type>("rgb_depth")}
+	{
+		cfg.disable_all_streams();
+		cfg.enable_stream(RS2_STREAM_ACCEL, RS2_FORMAT_MOTION_XYZ32F, ACCEL_RATE); // adjustable to 0, 63 (default), 250 hz
+		cfg.enable_stream(RS2_STREAM_GYRO, RS2_FORMAT_MOTION_XYZ32F, GYRO_RATE); // adjustable set to 0, 200 (default), 400 hz
+		cfg.enable_stream(RS2_STREAM_INFRARED, 1, IMAGE_WIDTH, IMAGE_HEIGHT, RS2_FORMAT_Y8, FPS);
+		cfg.enable_stream(RS2_STREAM_INFRARED, 2, IMAGE_WIDTH, IMAGE_HEIGHT, RS2_FORMAT_Y8, FPS);
+		cfg.enable_stream(RS2_STREAM_COLOR, IMAGE_WIDTH, IMAGE_HEIGHT, RS2_FORMAT_BGR8, FPS);
+		cfg.enable_stream(RS2_STREAM_DEPTH, IMAGE_WIDTH, IMAGE_HEIGHT, RS2_FORMAT_Z16, FPS);
+		profiles = pipe.start(cfg, [&](const rs2::frame& frame) { this->callback(frame); });
+	}
+
+	void callback(const rs2::frame& frame)
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+		// This lock guarantees that concurrent invocations of `callback` are serialized.
+		// Even if the API does not invoke `callback` in parallel, this is still important for the memory-model.
+		// Without this lock, prior invocations of `callback` are not necessarily "happens-before" ordered, so accessing persistent variables constitutes a data-race, which is undefined behavior in the C++ memory model.
+
+       	if (auto fs = frame.as<rs2::frameset>()) {
+            rs2::video_frame ir_frame_left = fs.get_infrared_frame(1);
+            rs2::video_frame ir_frame_right = fs.get_infrared_frame(2);
+            rs2::video_frame depth_frame = fs.get_depth_frame();
+            rs2::video_frame rgb_frame = fs.get_color_frame();
+            cv::Mat ir_left = cv::Mat(cv::Size(IMAGE_WIDTH, IMAGE_HEIGHT), CV_8UC1, (void*)ir_frame_left.get_data());
+			cv::Mat ir_right = cv::Mat(cv::Size(IMAGE_WIDTH, IMAGE_HEIGHT), CV_8UC1, (void *)ir_frame_right.get_data());
+			cv::Mat rgb = cv::Mat(cv::Size(IMAGE_WIDTH, IMAGE_HEIGHT), CV_8UC3, (void *)rgb_frame.get_data());
+			cv::Mat depth = cv::Mat(cv::Size(IMAGE_WIDTH, IMAGE_HEIGHT), CV_16UC1, (void *)depth_frame.get_data());
+			cv::Mat converted_depth;
+			float depth_scale = pipe.get_active_profile().get_device().first<rs2::depth_sensor>().get_depth_scale(); // for converting measurements into millimeters
+			depth.convertTo(converted_depth, CV_32FC1, depth_scale * 1000.f);
+			cam_type_ = cam_type{
+				new cv::Mat{ir_left},
+				new cv::Mat{ir_right},
```

Comment:
This is incorrect. The IR images have IR dots in them, and so cannot be used for SLAM. It is just dawning on me that none of the RealSense depth cameras would have stereo RGB. I guess we'll need the T265 after all.

Don't grab IR images for the moment. This PR, for now, should just be about RGB+Depth for scene reconstruction. At the same time, we'll need the stereo and IMU code in the future when the new camera arrives. So I think the way to do this is to check the RealSense camera model. If it is depth, only publish RGB+Depth. If it is tracking, only publish stereo-RGB+IMU.

Reply:
It seems like there is a way to turn off the emitter according to this: https://github.com/IntelRealSense/librealsense/pull/3066.
The `realsense-viewer` also has an option to turn off the emitter as well.

If we can turn off the emitter, the images look like regular grayscale images.

Reply:
Oh, very interesting. Give it a shot and see how it looks!